sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -131,16 +125,17 @@ Scheduler &Scheduler::getInstance() {
 }
 
 std::vector<EventImplPtr> Scheduler::getWaitList(EventImplPtr Event) {
-  std::lock_guard<std::mutex> lock(MGraphLock);
+  std::shared_lock<std::shared_timed_mutex> Lock(SMGraphLock);
   return GraphProcessor::getWaitList(std::move(Event));
 }
 
 void Scheduler::waitForEvent(EventImplPtr Event) {
+  std::shared_lock<std::shared_timed_mutex> Lock(SMGraphLock);
   GraphProcessor::waitForEvent(std::move(Event));
 }
 
 void Scheduler::cleanupFinishedCommands(EventImplPtr FinishedEvent) {
-  std::lock_guard<std::mutex> lock(MGraphLock);
+  std::unique_lock<std::shared_timed_mutex> Lock(SMGraphLock, std::try_to_lock);
```

Comment:
Why only try to lock?
There will be failure if the lock isn't taken, isn't it?
I'm not sure that behaviour of `shared_timed_mutex` has any similiarity to recursive mutex if `cleanupFinishedCommands` is involved in recursion.

Reply:
> Why only try to lock?

So that we can avoid a deadlock situation, where one thread is in the process of enqueueing (with a locked mutex) a currently blocked task that waits for another thread which is stuck at attempting cleanup.

> There will be failure if the lock isn't taken, isn't it?

+1, we need to check if the thread actually owns the lock below.

Reply:
```try_lock``` is used here because ```Scheduler::cleanupFinishedCommands``` is optional step - we can do cleanup later when a thread will own the mutex (the latest - when host accessor is destroyed).  
In addition, we use ```try_lock``` instead of ```lock``` here because in case of ```lock``` we will get a deadlock - one thread will wait if enqueue status is blocking in ```Command::enqueue``` func, another thread will try to lock the mutex in ```Scheduler::cleanupFinishedCommands``` func.

Reply:
>There will be failure if the lock isn't taken, isn't it?

@s-kanaev did you mean the following?

> Tries to lock the associated mutex without blocking by calling m.try_lock(). The behavior is undefined if the current thread already owns the mutex except when the mutex is recursive.

https://en.cppreference.com/w/cpp/thread/unique_lock/unique_lock

Reply:
> I'm not sure that behaviour of shared_timed_mutex has any similiarity to recursive mutex if cleanupFinishedCommands is involved in recursion.

I meant UB by this.


> There will be failure if the lock isn't taken, isn't it?

Here I think that it's never a good idea to continue working with `try_lock()` returning `false` where we are required to own a lock as it'll lead to writing to or reading from some wrong block of memory.

Reply:
Oh, I see, there is `Lock.owns_lock()` which is pretty nice to see.